File: Codex/Python/2025/Day2_Solution_2025.py
def ceil_div(a: int, b: int) -> int:
    return -(-a // b)


def range_sum(start: int, end: int) -> int:
    return (start + end) * (end - start + 1) // 2


def sum_invalid_in_range(start: int, end: int) -> int:
    total = 0
    max_digits = len(str(end))

    for half_len in range(1, max_digits // 2 + 1):
        base = 10 ** half_len + 1
        min_half = 1 if half_len == 1 else 10 ** (half_len - 1)
        max_half = 10 ** half_len - 1

        first = max(min_half, ceil_div(start, base))
        last = min(max_half, end // base)

        if first <= last:
            total += base * range_sum(first, last)

    return total


def main() -> None:
    with open("Day2_Input.txt", "r", encoding="utf-8") as file:
        content = file.read().strip()

    total = 0
    for part in content.split(","):
        if not part:
            continue
        start_text, end_text = part.split("-")
        total += sum_invalid_in_range(int(start_text), int(end_text))

    print(total)


if __name__ == "__main__":
    main()

File: Codex/Rust/2025/Day2_Solution_2025.rs
use std::fs;

fn parse_ranges(input: &str) -> Vec<(u64, u64)> {
    let mut ranges = Vec::new();

    for part in input.trim().split(',').filter(|s| !s.is_empty()) {
        let (start, end) = part
            .split_once('-')
            .expect("each range must contain a dash");
        let start: u64 = start.parse().expect("invalid range start");
        let end: u64 = end.parse().expect("invalid range end");
        ranges.push((start, end));
    }

    ranges
}

fn merge_ranges(mut ranges: Vec<(u64, u64)>) -> Vec<(u64, u64)> {
    if ranges.is_empty() {
        return ranges;
    }

    ranges.sort_unstable();

    let mut merged = Vec::with_capacity(ranges.len());
    let mut current = ranges[0];

    for &(start, end) in &ranges[1..] {
        if start <= current.1.saturating_add(1) {
            current.1 = current.1.max(end);
        } else {
            merged.push(current);
            current = (start, end);
        }
    }

    merged.push(current);
    merged
}

fn generate_invalid_ids(max_value: u64) -> Vec<u64> {
    let mut values = Vec::new();
    let mut power10 = 10_u64;

    for _half_len in 1..=10 {
        let start = power10 / 10;
        let end = power10 - 1;

        for half in start..=end {
            let value = half * power10 + half;
            if value > max_value {
                return values;
            }
            values.push(value);
        }

        if power10 > max_value {
            break;
        }
        power10 = match power10.checked_mul(10) {
            Some(next) => next,
            None => break,
        };
    }

    values
}

fn main() {
    let input = fs::read_to_string("Day2_Input.txt").expect("failed to read Day2_Input.txt");
    let ranges = parse_ranges(&input);
    let merged_ranges = merge_ranges(ranges);
    let max_end = merged_ranges.iter().map(|&(_, end)| end).max().unwrap_or(0);
    let invalid_ids = generate_invalid_ids(max_end);

    let mut sum: u128 = 0;
    let mut idx = 0usize;

    for (start, end) in merged_ranges {
        while idx < invalid_ids.len() && invalid_ids[idx] < start {
            idx += 1;
        }

        while idx < invalid_ids.len() && invalid_ids[idx] <= end {
            sum += invalid_ids[idx] as u128;
            idx += 1;
        }
    }

    println!("{sum}");
}
